Add build-loi command for updating an LoI file from <figure>s

apasel422 opened the pull request:
Text from the `<figcaption>`, if any, is preferred over that from the `<img>`'s `alt` attribute, though this can be controlled on a per-ID basis. If the resulting text is empty, `default_link_text` is used.

If you prefer alternative behavior, particularly around what to do if a `<figcaption>` or `alt` text is empty, please let me know and I'm happy to change it.

acabal commented:
Good idea. Did you run this on all of our books with LoIs to make sure it outputs the same thing already have? (Or makes appropriate corrections?)

robinwhittleton commented:
This could do with completions being added too.

apasel422 commented:
> Good idea. Did you run this on all of our books with LoIs to make sure it outputs the same thing already have? (Or makes appropriate corrections?)

I have not, but can try to do so today.

> This could do with completions being added too.

Will do.

apasel422 commented:
https://github.com/standardebooks/tools#running-commands-on-the-entire-corpus shows how to run a command on the entire corpus, but it assumes that you've already downloaded it. What's the recommended way to do so? Maybe the README can be updated to show this.

apasel422 commented:
As an example run, on https://github.com/standardebooks/lewis-carroll_alices-adventures-in-wonderland_john-tenniel this changes all the seemingly unique LoI link text to that of the `alt` attributes:

Compare [the frontispiece illustration `alt` text](https://github.com/standardebooks/lewis-carroll_alices-adventures-in-wonderland_john-tenniel/blob/c0b8d41211f5ce41884f4a9958958b917182804c/src/epub/text/frontispiece.xhtml#L11) to [the corresponding LoI text](https://github.com/standardebooks/lewis-carroll_alices-adventures-in-wonderland_john-tenniel/blob/c0b8d41211f5ce41884f4a9958958b917182804c/src/epub/text/loi.xhtml#L14).

acabal commented:
> https://github.com/standardebooks/tools#running-commands-on-the-entire-corpus shows how to run a command on the entire corpus, but it assumes that you've already downloaded it. What's the recommended way to do so? Maybe the README can be updated to show this.

You can use this script to download the corpus: https://github.com/standardebooks/web/blob/master/scripts/sync-ebooks

I haven't looked at it in a long time but it should still work fine.

acabal commented:
For Alice, I think that the LoI is just incorrect. It looks like those might have been captions from some other edition that were used instead of the alt text. `s-041` should definitely be fixed here.

apasel422 commented:
> `s-041` should definitely be fixed here.

Working on a fix, though I'm running into some unexpected behavior with the way lint tests work with files that are not present in the draft. Will post a PR shortly.

apasel422 commented:
In the current version of the command, 45 of the 81 ebook repos with an `loi.xhtml` file were initially changed after running `build-loi`. However, as will be seen, this was an overestimate. At a cursory glance, a few things stand out:

1. Some LoI text is not just text, but markup. [Here's an example](https://github.com/standardebooks/thomas-hardy_jude-the-obscure/blob/7457cf1fc7fe4df0d46ac9f6861976d14dfbe4a7/src/epub/text/loi.xhtml#L14). [The corresponding `<figure>`](https://github.com/standardebooks/thomas-hardy_jude-the-obscure/blob/7457cf1fc7fe4df0d46ac9f6861976d14dfbe4a7/src/epub/text/chapter-1-11.xhtml#L53) has `alt` text (not a `<figcaption>`), which of course cannot contain markup, so in this case the existing LoI file is correct. There isn't really anything we can do about this, as only the producer can know that the `alt` original text contained semantically relevant abbreviations. That said, if there is markup in a `<figcaption>`, we should probably retain it. I will make that change, though some `<figcaption>`s have more than just inline markup. (I saw at least one with a `<table>` and another with a `<blockquote>`.)

2. We should probably expose an option to prefer `alt` text for all figures in a work, as some figures [are captioned like `Figure 3`](https://github.com/standardebooks/edgar-allan-poe_the-narrative-of-arthur-gordon-pym-of-nantucket/blob/773efaa13b988571abc2db5c4600f16ac5e70524/src/epub/text/chapter-23.xhtml#L20) (presumably the author's intention) but [are labeled in the LoI using the more descriptive `alt` text](https://github.com/standardebooks/edgar-allan-poe_the-narrative-of-arthur-gordon-pym-of-nantucket/blob/773efaa13b988571abc2db5c4600f16ac5e70524/src/epub/text/loi.xhtml#L14). Though I'm not sure if that goes against the manual of style.

3. A number of books seem to be using obsolete LoI text, like the aforementioned Alice.

In summary, much like `build-spine`, we may want to indicate that `build-loi` is a good starting point for new works once they have their `<figure>`s added and captioned, but that it is probably not something you want to run all the time due to its propensity to undo intentional changes on the part of the producer.

acabal commented:
I think all of these cases can be solved:

1. build-loi can first check if an LoI item exists at the corresponding place for a figure. Then compute the LoI's text contents excluding tags (I.e. `EasXmlElement.inner_text()`. If the inner text matches either the figure `alt` text or `<figcaption>`, then don't overwrite it.
2. We can add a rule: if the figcaption is exactly `/Figure \d+/` then use `alt` instead.

apasel422 commented:
How should we handle the case in which an LoI item does *not* exist, but the `<figcaption>` contains non-phrasing content, such as `<p>` and `<table>`?

acabal commented:
For now just use `inner_text`, and the producer will have to massage it manually. Cases with complex elements will always require manual intervention.

apasel422 commented:
> For now just use `inner_text`, and the producer will have to massage it manually. Cases with complex elements will always require manual intervention.

I wonder if we should just use the default link text if we encounter block elements instead, so that `s-041` will definitely be emitted and the producer will be notified to manually inspect it. `inner_text` itself seems like it won't be especially useful because text from complex markup (e.g. table cells) will end up being joined together without spaces. We might as well just have the producer manually inspect these instead of emitting what will probably be garbage.

acabal commented:
Another thought is that only illustrations are meant to appear in the LoI, so if a figcaption has block children, why not just use the alt textg?

apasel422 commented:
OK, I've updated this PR with:

1. Closer alignment with `build-toc` regarding the `--stdout` flag and supporting multiple directories.
2. Removed the `--default-link-text` and `--prefer-alt-text` tags in lieu of:
3. Retaining non-block markup from `<figcaption>` and preferring `alt` text when block markup is present.
4. Preferring `alt` text over `<figcaption>Figure N</figcaption>`.

Ideally `build-loi` would also be able to create the LoI file if it doesn't exist, and this is definitely possible, but I was running into a problem in the test infrastructure in which the newly generated LoI file wasn't being compared due to its absence from the manifest/spine.

I won't be able to work on this again until July 1, but please let me know if there are any other changes you'd like to see, particularly since I haven't had the opportunity to re-run this on the corpus.

acabal commented:
Looks very good. The only feedback I have is on style:

1. `se_epub.generate_loi()` should return a string representing the LoI, just like `se_epub.generate_toc()` does. Then the command will either output the string or write the string to a file. 
2. We shouldn't be passing around a dom object in the command, the command shouldn't care about that. Instead `se_epub.generate_loi()` should get the dom itself using `self`. If there is no dom, then `se_epub.generate_loi()` should throw an exception that the command can then print.

I do think it would be useful for this command to write a new LoI file if one doesn't exist yet, because unlike the ToC, there is no default LoI file. Couldn't you simply ignore whatever irrelevant errors occur in that test case? In other words, the golden file would contain "error: file not in manifest" (or whatever) and we would consider that a successful test.

apasel422 commented:
> Couldn't you simply ignore whatever irrelevant errors occur in that test case? In other words, the golden file would contain "error: file not in manifest" (or whatever) and we would consider that a successful test.

The issue isn't an error occurring, but rather the new `loi.xhtml` file in the test's `golden` directory not even being compared at all if it's not also present in the test's input directory, which seems like a bug in the test infrastructure. I've updated this PR with a commit showing that for you to inspect.

acabal commented:
> > Couldn't you simply ignore whatever irrelevant errors occur in that test case? In other words, the golden file would contain "error: file not in manifest" (or whatever) and we would consider that a successful test.
> 
> The issue isn't an error occurring, but rather the new `loi.xhtml` file in the test's `golden` directory not even being compared at all if it's not also present in the test's input directory, which seems like a bug in the test infrastructure. I've updated this PR with a commit showing that for you to inspect.

@vr8hub what do you think?

apasel422 commented:
Here's a minimal PR showing that these golden files are ignored: https://github.com/standardebooks/tools/pull/726

vr8hub commented:
The golden file is ignored because it doesn't exist in the draft. We have no commands that create a file out of whole cloth (other than build, and it's a special case), so there is currently no means to deal with this. There are one or three other problems to solve in the testing structure, but I chose to concentrate on the lint tests first, because that's where the greatest need was. I'll look at this, but my kids are still in town, so it will be a few days.

acabal commented:
OK, I'll merge this in for now because otherwise it looks good, and Vince if and when you have some time maybe you can take a look to see if we can update the test system to accommodate this case.

Thanks Andrew!

vr8hub commented:
Definitely, it will just be a couple of weeks before they leave and I can get to it.
